pkg/controllermanager/controller/controllerdeployment/controllerdeployment_control_test.go

```diff
@@ -0,0 +1,214 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controllerdeployment_test
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/golang/mock/gomock"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	fakeclientmap "github.com/gardener/gardener/pkg/client/kubernetes/clientmap/fake"
+	"github.com/gardener/gardener/pkg/client/kubernetes/clientmap/keys"
+	. "github.com/gardener/gardener/pkg/controllermanager/controller/controllerdeployment"
+	"github.com/gardener/gardener/pkg/logger"
+	mockclient "github.com/gardener/gardener/pkg/mock/controller-runtime/client"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+)
+
+var _ = Describe("Controller", func() {
```

Comment:
Should we rather have an envtest for this reconciler (similar to https://github.com/gardener/gardener/blob/master/pkg/controllermanager/controller/shoot/shoot_retry_control_test.go)?

Reply:
As discussed with @timebertt, @stoyanr and @ialidzhikov last week, adding envtests comes with a price, especially if we setup and tear down the environment in every suite and our pipeline resources aren't sufficient enough (sometimes) which has ended in flaky tests so far. The CC cache can improve the situation but we are not sure, yet.
Bottom line, I think the controller is so simple and very similar to https://github.com/gardener/gardener/blob/master/pkg/controllermanager/controller/controllerregistration/controllerregistration_control_test.go, so that these unit tests cover the given cases sufficiently and an envtest doesn't seem too beneficial to me (except for consistency reasons). WDYT?

Reply:
Yeah, I'm a bit unsure about this problem domain. While doing my recent refactorings in the controllers, I learned that it was very tedious to adapt the unit tests because too many implementation details were tested (all these mocks with many different clients, etc.). My impression is that at least the tests for the reconcilers/controllers would be easier to maintain and way clearer if we used envtests.
I don't insist on changing it now for this PR (everything is implemented already), but we should have a talk about it and clarify how to proceed in the future. If the only argument for not doing it is a potentially unstable CI/CD environment then let's try to work on this with the colleagues or even switch to a more stable env. But maybe you guys also discussed other reasons.

Reply:
> If the only argument for not doing it is a potentially unstable CI/CD environment

That's actually not my main concern but that we should evaluate if an integration test over unit test would be beneficial here. In this controller, we have only a few dependencies, so I wonder why we shouldn't prefer small and fast unit tests. But yes, let's have a discussion about this.

Reply:
> While doing my recent refactorings in the controllers, I learned that it was very tedious to adapt the unit tests because too many implementation details were tested (all these mocks with many different clients, etc.). My impression is that at least the tests for the reconcilers/controllers would be easier to maintain and way clearer if we used envtests.

+100, that's also the reason why I'm pushing envtests for our controllers. I learned to hate mocks for controller tests during my recent refactorings / cleanups as well. They force you to adapt soo many unit tests that rely on mocks when changing a simple helper func from `TryUpdate` to `Patch`. 

Yes, let's have a chat about this :)